### DISCUSSION TITLE
Preconditioner bilinear form of examples 22 and 25

### DISCUSSION THREAD

**Discussion by mdeuse** [mfem/mfem#2560] [Q&A] [answered]
Hi,

Could someone explain to me how were the bilinear forms of the preconditioner of examples 22 and 25 selected? Is there any documentation/literature that would explain the reasons behind these specific choice of bilinear forms?

Thank you in advance,

Mathieu

**Comment by tzanio** [comment]
ping: @mlstowell, @psocratis

**Reply by mlstowell** [reply]
This is a good question. Unfortunately, I'm not aware of any literature on this although it may well exist. As I recall I stumbled upon this operator while trying to approximate the absolute value of the operator needed in the H(curl) case. This was motivated by the fact that the iterative solver required a positive definite preconditioner. Socratis (@psocratis), a real expert in such matters, then suggested other improvements which, if I remember correctly, drastically reduced the number of iterations in the other two cases.

It would not surprise me to learn that these choices depend on the relative size of the real and imaginary parts of the mass operator. I'm sorry I don't have anything more substantial to offer.

**Reply by mdeuse** [reply]
Thanks for the reply. 

To give you some context, The problem that I solve is quite close to examples 22 and 25 (in my case it is a convected wave equation, with a scalar H1 field). As I am trying to improve the convergence speed in the case of an iterative solver, I am investigating different forms of preconditioners. One type of preconditioner that I would like to implement employs complex coefficients. However, if I understand correctly, examples 22 and 25 are only suited to preconditioners with real coefficients.

So far, I have not been able to adapt the code of these examples to complex coefficients. Is there a way to use a SesquilinearForm instead of a BilinearForm for the preconditioner? Or manually, to have a block operator which inculdes off-diagonal terms to account for the imaginary part?

Thanks again

**Comment by psocratis** [comment, marked as answer]
Hi @mdeuse, 

As you may know, preconditioning time-harmonic wave operators is an open research area. The truth is that for moderate size problems (especially in 2D) the best option is to use a direct solver. Iterative solvers/preconditioners for these problems are very hard to construct since the system is indefinite and their convergence depends strongly on the frequency. 

The preconditioners in examples 22/25 are indeed very basic and are not intended to be used for high frequency or really large problems. In some way, they are based on the idea of the "Shifted Laplacian" preconditioner (a very basic one). A more sophisticated preconditioner is implemented [here](https://github.com/mfem/mfem/blob/DST-maxwell-solver-dev/examples/maxwell-solver-dev/helmholtzp.cpp) based on the "Source Transfer Preconditioner" for Helmholtz and Maxwell. This preconditioner is very effective, however, there are limitations (geometry/mesh/BCs).

Back to your question ... 
The two approaches (using sesquilinear forms or manually constructing the 2x2 blocks) are essentially the same. Using  `Assemble` and `FormLinearSystem` for a `SesquiliearForm` does exactly the same thing underneath, i.e., it forms the 2x2 real system. This is needed for all the mfem solvers (GMRES, CG, etc.) Of course, in examples 22/25, using preconditioners which are constructed just from real operators it was much easier since we were able to use the `BlockDiagonal` preconditioner out of the box. I am sure it will be relatively straightforward to derive/adjust this for your purpose and build preconditioners with off-diagonal terms as well. 

My apologies for not being able to help any further right now. It might help if you share with us the preconditioner you plan to construct. 

Best, 

Socratis

**Reply by mdeuse** [reply]
Hi Socratis,

Thank you for your detailed reply. I have already integrated the MUMPS direct solver to my project, but as you mention it is not really suitable as the size of the problem grows, which is precisely why I am investigating iterative solvers.

Indeed, I have been looking into the Shifted Laplacian Preconditioner. In my initial question, I was essentially wondering about the contribution of the loss coefficient. I have also read about some formulations of this preconditioner that employ a complex-valued mass coefficient, hence my second question. So I will try to replace the block diagonal preconditioner by a 2x2 system, as you mentioned.

I did not know about the Source Transfer Preconditioner, thanks for the suggestion, that might surely be useful here.

Thanks for the help!

Best,

Mathieu
